Migration guide from k8s-workload-registrar (#40)
spiffe/spire-controller-manager: migration/README.md

# Kubernetes Workload Registrar Migration

## Introduction

> This guide will walk you through you through how to migrate an existing Kubernetes Workload Registrar deployment to SPIRE Controller Manager. Existing entries created by the Kubernetes Workload Registrar aren't compatible with SPIRE Controller Manager so they'll be deleted and replaced with new entries. Workloads will continue to function with the old entries until their certificates expire, after which they'll get new certificates based on the new entries.

Reviewer: ```suggestion
This guide will walk you through how to migrate an existing Kubernetes Workload Registrar deployment to SPIRE Controller Manager. Existing entries created by the Kubernetes Workload Registrar aren't compatible with SPIRE Controller Manager so they'll be deleted and replaced with new entries. Workloads will continue to function with the old entries until their certificates expire, after which they'll get new certificates based on the new entries.
```

> **Note**
> As we'll be deleting and creating entries, it's important to do this migration during a downtime window.

## Clean up Kubernetes Workload Registrar Resources

First we need to clean up the Kubernetes Workload Registrar and its resources.

1. (CRD mode only) Delete the SpiffeId Custom Resource Definition (CRD). This will delete all entries created by the k8s-workload-registrar. 
   ```shell
   kubectl delete crd spiffeids.spiffeid.spiffe.io
   ```

> 1. Delete the `ValidatingWebhookConfiguration`, `Service`, `Roles`, and other k8s-workload-registrar config. Not all of the resources below are applicable for all k8s-workload-registrar modes, so if there's a "not found" message it's safe to ignore. In genreal make sure to clean up any Kubernetes Workload Registrar resources aside from the SPIRE Server and Kubernetes Worklaod Registrar itself. Those will be removed below.

Reviewer: ```suggestion
1. Delete the `ValidatingWebhookConfiguration`, `Service`, `Roles`, and other k8s-workload-registrar config. Not all of the resources below are applicable for all k8s-workload-registrar modes, so if there's a "not found" message it's safe to ignore. In general make sure to clean up any Kubernetes Workload Registrar resources aside from the SPIRE Server and Kubernetes Workload Registrar itself. Those will be removed below.
```

                 -f config/role_binding.yaml \
                 -f config/spire-controller-manager-config.yaml \
                 -f config/spire-server.yaml
   ```

1. Create the `ClusterSpiffeId` custom resource. The below example will create SPIFFE IDs with this shape: `spiffe://<trust domain>/ns/<namespace>/sa/<serviceaccount>`. Only Pods with the label `spiffe.io/spiffe-id: true` will have entries auto-created. This corresponds to the `identity_template` and `identity_template_label` configurables from CRD mode Kubernetes Workload Registrar. 
   ```shell
   kubectl apply -f config/clusterspiffeid.yaml
   ```

   > **Note**
   > See [FAQs](#faqs) or instructions on how to translate [label](#how-do-i-do-label-based-workload-registration), [annotation](#how-do-i-do-annotation-based-workload-registration), and [service account](#how-do-i-do-service-account-based-workload-registration) based workload registration. Also see [ClusterSPIFFEID defintion](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) for more information on how to create the shape most suitable to your environment.

## Verify Spire Controller Manager Deployment

> First verify the Pods came up correctly. The `spire-server-0` Pod should have two containers running in it.

Reviewer 2: This will only be the case if you have edited your previous pod to remove the registrar container. Might be obvious, but also might be worth mentioning that as an explicit step.

Reviewer 2: Actually, this should be obvious from the contents of the statefulset YAML.

  name: federation
spec:
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  podSelector:
    matchLabels:
      spiffe.io/spiffe-id: "true"
  dnsNameTemplates: ["{{ .PodMeta.Name }}", "my-custom-dns-name"]

```

### Does SPIRE Controller Manager automatically populate DNS Names of Services a Pod is attached to?

SPIRE Controller Manager doesn't monitor Endpoints like Kubernetes Workload Registrar did, so it won't do this automatically. A workaround is to use the `app` label to populate DNS Names using `dnsNameTemplates` field of the [ClusterSPIFFEID CRD](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md), assuming you are using `app` as your selector and it matches the name of the `Service`.

```yaml
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: federation
spec:
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  podSelector:
    matchLabels:
      spiffe.io/spiffe-id: "true"
  dnsNameTemplates: ["{{ index .PodMeta.Labels \"app\" }}.{{ .PodMeta.Namespace }}.svc.cluster.local"]

```

> If you require these DNS Names to be automatically populated, please open a [new issue](https://github.com/spiffe/spire-controller-manager/issues/new) with your use case.

Reviewer 3: Issue created for this: https://github.com/spiffe/spire-controller-manager/issues/48

Author: Updated the readme with this ticket. Is this a showstopper for migrating nginx service mesh?

Reviewer 3: Yes, due to the unpredictability of how a user may choose to label their Pods. I also created a second blocking issue that was discovered while testing, https://github.com/spiffe/spire-controller-manager/issues/49, though is unrelated to this guide.

Reviewer: Bump into the same and commented on the issue to give my 50 cents.

### Can SPIRE Controller Manager be deployed in a different Pod from SPIRE Server?

This is not supported with SPIRE Controller Manager, they must by in the same Pod. If you require them to be in seperate Pods, please open a [new issue](https://github.com/spiffe/spire-controller-manager/issues/new) with your use case.

### How do i see SPIRE Controller Manager logs?

```shell
$ kubectl logs spire-server-0 -n spire -c spire-controller-manager
2022-12-13T00:41:21.362Z	INFO	setup	Config loaded	{"cluster name": "demo-cluster", "trust domain": "example.org", "ignore namespaces": ["kube-system", "kube-public", "istio-system", "spire", "local-path-storage"], "gc interval": "10s", "spire server socket path": "/spire-server/api.sock"}
2022-12-13T00:41:21.764Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8082"}
2022-12-13T00:41:21.807Z	INFO	webhook-manager	Minting webhook certificate	{"reason": "initializing", "dnsNames": ["spire-controller-manager-webhook-service.spire.svc"]}
2022-12-13T00:41:21.828Z	INFO	webhook-manager	Minted webhook certificate
2022-12-13T00:41:21.844Z	INFO	webhook-manager	Webhook configuration patched with CABundle
```

> ### I'm using CRD mode Kubernetes Workload Registrar and it gets stuck deleting the SpiffeId CRD. What do I do?

Reviewer 3: Since the new CRDs also appear to have finalizers, could we run into the same issue with those as we did with the SpiffeID CRDs?

Author: I see finalizer mentioned in the repo but I dont see it being attached to any resources as it was the k8s-workload-registrar. So I dont think you'll run into that issue. @azdagron can you confirm?

Reviewer 3: I wasn't sure, because I saw it in the RBAC permissions.

Reviewer 4: Those were added by default with kubebuilder. We don't currently make use of finalizers.

# Kubernetes Workload Registrar Migration

## Introduction

> This guide will walk you through how to migrate an existing Kubernetes Workload Registrar deployment to SPIRE Controller Manager. Existing entries created by the Kubernetes Workload Registrar aren't compatible with SPIRE Controller Manager so they'll be deleted and replaced with new entries. Workloads will continue to function with the old entries until their certificates expire, after which they'll get new certificates based on the new entries.

Author: updated

> **Note**
> As we'll be deleting and creating entries, it's important to do this migration during a downtime window.

> ## Clean up Kubernetes Workload Registrar Resources

Author: Thats going to be tricky to have them running side by side. Perhaps we can just strip the finalizers by patching the CRDs after the registrar is removed?

> ## Clean up Kubernetes Workload Registrar Resources

Author: This script does the trick:

```
for ns in $(kubectl get ns | awk '{print $1}' | tail -n +2)
do
  if [ $(kubectl get spiffeids -n $ns 2>/dev/null | wc -l) -ne 0 ]
  then
    kubectl patch spiffeid $(kubectl get spiffeids -n $ns | awk '{print $1}' | tail -n +2) --type='merge' -p '{"metadata":{"finalizers":null}}' -n $ns
  fi
```

> Allowing the value of labels to directly populate a SPIFFE ID gives the power to create arbitrary SPIFFE IDs to anyone that can deploy a Pod in your cluster. It's better to define a SPIFFE ID using a template that doesn't depend on a label. See [ClusterSPIFFEID defintion](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) for more information. 

### How do I do annotation based workload registration?

There is no equivalent to this configuration in SPIRE Controller Manager. Annotations specifically don't allow for selecting Pods with a specific annotation, which SPIRE Controller Manager relies on. The easiest path forward is to convert the annotations to labels and use [label based workload registration]((#how-do-i-do-label-based-workload-registration)).

### How do I do service account based workload registration?

This can be done with the SPIRE Controller Manager with a config like the below:

```yaml
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: service-account-based
spec:
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
```

> **Note**

> > This will create entries for every Pod in the system. Its better to restrict it with a label like in the main example in `config/clusterspiffeid.yaml`. Also see [ClusterSPIFFEID defintion](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) for more information.

Reviewer 4: > Its better to restrict it

I think possibly a better approach would be to just explain the pros and cons and let folks decide for themselves which is better according to their circumstances. There are valid scenarios where you don't need to restrict by label.

Author: Updated note.

### How do I federate trust domains?

> With Kubernetes Workload Regisrar the Pod annotation `spiffe.io/federatesWith` is used to create SPIFFE ID's that federate with other trust domains. For example:

Reviewer 4: typo: Registrar

```

### Does SPIRE Controller Manager automatically populate DNS Names of Services a Pod is attached to?

SPIRE Controller Manager doesn't monitor Endpoints like Kubernetes Workload Registrar did, so it won't do this automatically. A workaround is to use the `app` label to populate DNS Names using `dnsNameTemplates` field of the [ClusterSPIFFEID CRD](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md), assuming you are using `app` as your selector and it matches the name of the `Service`.

```yaml
apiVersion: spire.spiffe.io/v1alpha1
kind: ClusterSPIFFEID
metadata:
  name: federation
spec:
  spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
  podSelector:
    matchLabels:
      spiffe.io/spiffe-id: "true"
  dnsNameTemplates: ["{{ index .PodMeta.Labels \"app\" }}.{{ .PodMeta.Namespace }}.svc.cluster.local"]

```

If you require these DNS Names to be automatically populated, please update [#48](https://github.com/spiffe/spire-controller-manager/issues/48) with your use case.

### Can SPIRE Controller Manager be deployed in a different Pod from SPIRE Server?

> This is not supported with SPIRE Controller Manager, they must by in the same Pod. If you require them to be in seperate Pods, please open a [new issue](https://github.com/spiffe/spire-controller-manager/issues/new) with your use case.

Reviewer 4: typo: ...they must be...

2022-12-13T00:41:21.807Z	INFO	webhook-manager	Minting webhook certificate	{"reason": "initializing", "dnsNames": ["spire-controller-manager-webhook-service.spire.svc"]}
2022-12-13T00:41:21.828Z	INFO	webhook-manager	Minted webhook certificate
2022-12-13T00:41:21.844Z	INFO	webhook-manager	Webhook configuration patched with CABundle
```

### I'm using CRD mode Kubernetes Workload Registrar and it gets stuck deleting the SpiffeId CRD. What do I do?

This can happen if the Kubernetes Workload Registrar is deleted before all the SpiffeId custom resources are removed. To get around this, manually remove the finalizers with the below script and try deleting the CRD again.

```shell
for ns in $(kubectl get ns | awk '{print $1}' | tail -n +2)
do
  if [ $(kubectl get spiffeids -n $ns 2>/dev/null | wc -l) -ne 0 ]
  then
    kubectl patch spiffeid $(kubectl get spiffeids -n $ns | awk '{print $1}' | tail -n +2) --type='merge' -p '{"metadata":{"finalizers":null}}' -n $ns
  fi
done
```

### Why can't Kubernetes Workload Registrar entries be reused with SPIRE Controller Manager?

> SPIRE Controller Manager uses a different scheme for parenting SPIFFE IDs. Though it is technically possible to modify all the entries, its a lot easier to just allow SPIRE Controller Maanger to automatically replace the entries.

Reviewer 4: typo: Manager

This is not supported with SPIRE Controller Manager, they must by in the same Pod. If you require them to be in seperate Pods, please open a [new issue](https://github.com/spiffe/spire-controller-manager/issues/new) with your use case.

> ### Can I manually create entries like I could with the CRD Kubernetes Worklaod Registrar?

Reviewer: ```suggestion
### Can I manually create entries like I could with the CRD Kubernetes Workload Registrar?
```
